pahfit/model.py

```diff
@@ -0,0 +1,581 @@
+    STORES is a description of the physics: what features are there and
+    what are their properties, regardless of the instrument with which
+    those features are observed. The methods provided by this class,
+    form the connection between those physics, and what is observed.
+    During fitting and plotting, those physics are converted into a
+    model for the observation, by applying instrumental parameters from
+    the instrument.py module.
+
+    The main thing that defines a model, is the features table, loaded
+    from a YAML file given to the constructor. After construction, the
+    Model can be edited by accessing the stored features table directly.
+    Changing numbers in this table, is allowed, and the updated numbers
+    will be reflected when the next fit or initial guess happens. At the
+    end of these actions, the fit or guess results are stored in the
+    same table.
+
+    The model can be saved.
+
+    The model can be copied.
+
+    Attributes
+    ----------
+    features : Features
+        Instance of the Features class. Can be edited on-the-fly.
+        Non-breaking behavior by the user is expected. Changes will be
+        reflected at the next fit, guess, or plot call.
+
+    """
+
+    def __init__(self, features: Features):
+        """
+        Parameters
+        ----------
+        features: Features
+            Features table.
+
+        """
+        if len(features) < 2:
+            raise PAHFITModelError(
+                "Fewer than 2 features! Single component models are no allowed!"
+            )
+
+        self.features = features
+        # if set to False, use set fwhm for lines to value in features
+        # table at model construction
+        self.use_instrument_fwhm = True
+
+        # store fit_info dict of last fit
+        self.fit_info = None
+
+    @classmethod
+    def from_yaml(cls, pack_file):
+        """
+        Generate feature table from YAML file.
+
+        Parameters
+        ----------
+        pack_file : str
+            Path to YAML file, or name of one of the default YAML files.
+
+        Returns
+        -------
+        Model instance
+
+        """
+        path = find_packfile(pack_file)
+        features = Features.read(path)
+        return cls(features)
+
+    @classmethod
+    def from_saved(cls, saved_model_file):
+        """
+        Parameters
+        ----------
+        saved_model_file : str
+           Path to file generated by Model.save()
+
+        Returns
+        -------
+        Model instance
+        """
+        # features.read automatically switches to astropy table reader.
+        # Maybe needs to be more advanced here in the future.
+        features = Features.read(saved_model_file, format="ascii.ecsv")
+        return cls(features)
+
+    def save(self, fn, **write_kwargs):
+        """Save the model to disk.
+
+        Only ECSV supported for now. Models saved this way can be read
```

Comment:
I did a quick experiment, and saved fit.fit_info into the metadata (which is supposed to contain the covariance matrix if it's actually returned). This adds several thousand lines of binary blob data to the table header, like in this snippet

```
...
meta: !!omap
# - {unit: null}
# - {use_instrument_fwhm: true}
# - fit_info:
#     cov_x: null
#     fjac: !numpy.ndarray
#       buffer: !!binary |
#         ckNiY3I3MUNpY0lkZzhuUDJLdkVQNFpRYWxkbVJzUS9jd2hPa1pwVHdULzJqcHNOMC9teVA4MUZ0
#         ei81cXJNL280SE5hYm1Hd3o5RFMzOHY1MU80UHpUZFJuRXRyN00vWUo2dmg2em5zRC9UR1pQZHFB
#         YXdQNWgvM1NzZ29xVS9mQThKTWJDVm9EK2xKTUpKYkFuRFAxRDNCYmRtRmI0L2R5MU9XUmt1MEQ5
#         WkdUTlNhcS9HUDVaQ2l4Ym1mTDQvblllUWxxdTJ6ait0eEo4T1hsdkZQeTVlNmFNbEJMOC9zWkwx
...
```

It round trips just fine.

Reply:
And of course instrument and redshift used to be saved, but we will reinstate that in the future.

Reply:
Huh.  I thought if you do "normal arrays" it will put them in ascii.  Encoded-binary blob is fine too I guess (but people better not hand-edit those things).   ASDF will be much faster than doing uuencoding.  But probably not really important except doing thousands of spectra in a row (which some crazies will do).

Reply:
Sorry why are instrument and redshift not being saved?

Reply:
I was still working under the paradigm that 'instrument' is a property of spec. But as suggested in the above comment, we can make the "instrument marriage" happen at the first call to fit (and later refactor to have it at construction). For now, I'll do this by putting instrument in features.meta.

Reply:
Yes: either "arranged marriage on instantiation" or "fall in love when you see your first spectrum".  We'll have to decide what to do if the Spectrum/a passed do not agree with the pre-configured (list of) instruments.  In terms of refactoring (after dev merge), also see (and be thinking about) #250.

Reply:
marriage = "instrument and redshift saved into Model".  Possibly the (set of) spectra instrument(s) + redshift should just override the pre-existing values, with a warning.